### PR TITLE
Make conn_id unique in Connections table

### LINKED ISSUE #8608: Make conn_id unique in Connections table
s we discussed in the thread on Spec thread for our new API, the idea
came up of making conn_id unique in Airflow 

<https://lists.apache.org/thread.html/rfb6f7d95b2754fda9dd09b08444214dfad12d10f143d32de0fcf4104%40%3Cdev.airflow.apache.org%3E>

The people in that thread seemed to be in favour of it:

- It is often confusing to users (there have been a few bug reports about it over the years)
- It's questionable if it actually works well or not
- There are better/smarter tools for loadbalancing connections to a DB than picking one of a random list
- For Hive at least it has been implemented another way - allowing two host's in a single connection string https://github.com/apache/airflow/pull/4708 (available in 1.10.6)
- It makes the HTTP API confusing (needing a integer connection ID, and a string conn_id field)

Given the downsides/work arounds, and the confusion this causes we should remove this (mis)feature from Airflow 2.0.

We should take some care in the migration that does this to capture the unique constraint error and report it a a meaninful way to the user.

### PR DESCRIPTION
**Pull request by Hasan-J**
Closes #8608 

Added an alembic revision, if the user already has multiple records with the same `conn_id`
in the connections table, it throws an exception during `airflow db upgrade` with a clear
message.

---
Make sure to mark the boxes below before creating PR: [x]

- [x] Description above provides context of the change
- [x] Unit tests coverage for changes (not needed for documentation changes)
- [x] Target Github ISSUE in description if exists
- [x] Commits follow "[How to write a good git commit message](http://chris.beams.io/posts/git-commit/)"
- [x] Relevant documentation is updated including usage instructions.
- [x] I will engage committers as explained in [Contribution Workflow Example](https://github.com/apache/airflow/blob/master/CONTRIBUTING.rst#contribution-workflow-example).

---
In case of fundamental code change, Airflow Improvement Proposal ([AIP](https://cwiki.apache.org/confluence/display/AIRFLOW/Airflow+Improvements+Proposals)) is needed.
In case of a new dependency, check compliance with the [ASF 3rd Party License Policy](https://www.apache.org/legal/resolved.html#category-x).
In case of backwards incompatible changes please leave a note in [UPDATING.md](https://github.com/apache/airflow/blob/master/UPDATING.md).
Read the [Pull Request Guidelines](https://github.com/apache/airflow/blob/master/CONTRIBUTING.rst#pull-request-guidelines) for more information.

### CODE REVIEW

File: airflow/migrations/versions/8d48763f6d53_add_unique_constraint_to_conn_id.py
Comment (xinbinhuang) [inline airflow/migrations/versions/8d48763f6d53_add_unique_constraint_to_conn_id.py]:
```suggestion
          
            with op.batch_alter_table('connection') as batch_op:
                batch_op.create_unique_constraint(
                    constraint_name="unique_conn_id",
                    columns=["conn_id"]
                )
```

You don't need to create your own branching logic. [Alembic](https://alembic.sqlalchemy.org/en/latest/ops.html#alembic.operations.Operations.batch_alter_table) will handle it for you internally.

**Review by potiuk — COMMENTED** [review]
I think we need a note in UPDATING.md about this change. Also (but probably later) some migration tool for airflow 2.0 should also deal with this case if someone already has connection in their database . We will have to figure out migration scheme, but I think a note in UPDATING.md will be enough to remember about doing it.

**Comment by Hasan-J** [comment]
I have removed the test method `test_connection_metastore_secrets_backend` in `tests/secrets/test_secrets_backends.py` based on the fact that it's using mutiple connections with the same `conn_id`. Maybe someone can confirm that this indeed should be removed !

File: tests/secrets/test_secrets_backends.py
Comment (feluelle) [inline tests/secrets/test_secrets_backends.py]: I think we still need this test but maybe modify it to test one connection instead of two.

**Comment by mik-laj** [comment]
I wonder if we should mark this field as non nullable. 
Here is the master table:
```
                                          Table "public.connection"
       Column       |          Type           | Collation | Nullable |                Default
--------------------+-------------------------+-----------+----------+----------------------------------------
 id                 | integer                 |           | not null | nextval('connection_id_seq'::regclass)
 conn_id            | character varying(250)  |           |          |
 conn_type          | character varying(500)  |           |          |
 host               | character varying(500)  |           |          |
 schema             | character varying(500)  |           |          |
 login              | character varying(500)  |           |          |
 password           | character varying(5000) |           |          |
 port               | integer                 |           |          |
 extra              | character varying(5000) |           |          |
 is_encrypted       | boolean                 |           |          |
 is_extra_encrypted | boolean                 |           |          |
Indexes:
    "connection_pkey" PRIMARY KEY, btree (id)
```
conn_type should also not be empty, but that's another issue.

**Comment by mik-laj** [comment]
Would it be a big problem to use this field as the primary key instead of just adding a second unique index?  The primary key is also a unique index.

File: airflow/models/connection.py
Comment (ashb) [inline airflow/models/connection.py]: This looks odd -- I would have expected it to be this
```suggestion
    __table_args__ = (
        UniqueConstraint(conn_id, name='unique_conn_id'),
    )
```

**Review by ashb — COMMENTED** [review]
I think the airflow.secrets.local_filesystem needs updating too -- it still supports multiple connections with the same id.

**Comment by Hasan-J** [comment]
> I think the airflow.secrets.local_filesystem needs updating too -- it still supports multiple connections with the same id.

Oh, this is my first time reading about this backend. Should we stop and raise an exception in case of multiple connections or just load that last one specified ?

File: airflow/models/connection.py
Comment (ashb) [reply airflow/models/connection.py]: Actually - even simpler is adding `unique=True` on Line 111: `conn_type = Column(String(500))`- ? `conn_type = Column(String(500), unique=True)`

**Comment by mik-laj** [comment]
@Hasan-J It would be a better idea to raise an exception.

File: airflow/models/connection.py
Comment (ashb) [inline airflow/models/connection.py]: Are these used anywhere? It doesn't appear to be, if so please don't add them in this PR.

Comment (ashb) [reply airflow/models/connection.py]: Other than this change this PR looks good to me.

File: UPDATING.md
Comment (ashb) [inline UPDATING.md]:
```suggestion
### Connections must now have a unique conn_id
```

File: UPDATING.md
Comment (ashb) [inline UPDATING.md]:
```suggestion
Previously, Airflow allowed users to add more than one connection with the same `conn_id` and on access it would choose one connection randomly. This acted as a basic load balancing and fault tolerance technique, when used in conjunction with retries.
```

File: UPDATING.md
Comment (ashb) [inline UPDATING.md]:
```suggestion
Now the `conn_id` will be unique. New users will not have to deal with any consequences. But for existing users that already have data in their metadata database, they will have to manage those duplicate connections before upgrading the database with `airflow db upgrade`.
```

File: airflow/models/connection.py
Comment (ashb) [inline airflow/models/connection.py]:
```suggestion
    conn_id = Column(String(ID_LEN), unique=True, nullable=False)
```

**Review by ashb — DISMISSED** [review]
A few small doc fixes (case, and correct quoting for markdown files)

**Comment by mik-laj** [comment]
I can see CI/CD is sad. Can you fix it?

**Review by mik-laj — APPROVED** [review]

**Review by feluelle — APPROVED** [review]

**Comment by mik-laj** [comment]
@Hasan-J I accepted and for some reason I cannot merge this change. Additionally, I have a small request: when you finish working on your change, please write a comment so that we will receive notifications about your activity.  If you do not make a comment, we do not see the progress of the work and we must remember to come back to this change.

**Comment by Hasan-J** [comment]
Got it, is there something I can do to help with troubleshooting?

**Comment by mik-laj** [comment]
> Got it, is there something I can do to help with troubleshooting?

@ashb  asked for changes to this ticket, but is now on vacation.  This blocks the merging of this change.  Could you close this PR and create a new one?

**Comment by potiuk** [comment]
I dismissed the review from @ashb now - we think all has been addressed here. So you only need to rebase it now @Hasan-J !

**Review by potiuk — APPROVED** [review]

File: UPDATING.md
Comment (kaxil) [inline UPDATING.md]:
```suggestion
Now the `conn_id` will be unique. If you already have duplicates in your metadata database, you will have to manage those duplicate connections before upgrading the database.
```

File: airflow/migrations/versions/8d48763f6d53_add_unique_constraint_to_conn_id.py
Comment (kaxil) [inline airflow/migrations/versions/8d48763f6d53_add_unique_constraint_to_conn_id.py]: This value can change in future.

So we should hard code it in the migration itself, instead of importing

**Comment by Hasan-J** [comment]
I think we're ready to merge this !

**Comment by potiuk** [comment]
Wohoo :tada: !

**Comment by kaxil** [comment]
Thanks @Hasan-J 🎉

**Comment by Hasan-J** [comment]
I really appreciate all of you. Thank you.

**Comment by mik-laj** [comment]
I thought about this change even longer and I think we need to make additional changes. The change to the database is not enough, but we also need to update the Python API.
https://github.com/apache/airflow/issues/10135

**Comment by Hasan-J** [comment]
I've also had my thoughts about this change, but didn't mind it because it "worked".
I will take a look and open a PR soon :)

**Comment by feluelle** [comment]
@Hasan-J please add a comment to the issue so we can assign it to you. :)

**Comment by mik-laj** [comment]
@Hasan-J Can you comment on the ticket? Without it, I can't assign you to him.

**Pull request merged**
